mango/express/api.py

```diff
@@ -0,0 +1,260 @@
+    evenly.
+
+    Example:
+    ```python
+    async with run_with_tcp(2, Agent(), Agent(), (Agent(), dict(aid="my_agent_id"))) as c:
+        # do your stuff
+    ```
+
+    Args:
+        num (int): number of tcp container
+        agents (args): list of agents which shall run
+
+    Returns:
+        RunWithTCPManager: the async context manager to run the agents with
+    """
+    return RunWithTCPManager(num, agents, addr=addr, codec=codec)
+
+
+def run_with_mqtt(
+    num: int,
+    *agents: tuple[Agent, dict],
+    broker_addr: tuple[str, int] = ("127.0.0.1", 1883),
+    codec: None | Codec = None,
+) -> RunWithMQTTManager:
+    """Create and return an async context manager, which can be used to run the given
+    agents in `num` automatically created mqtt container. The agents are distributed according
+    to the topic
+
+    Args:
+        num (int): _description_
+        agents (args): list of agents which shall run, it is possible to provide a tuple
+                       (Agent, dict), the dict supports "aid" for the suggested_aid and "topics" as list of topics the agent
+                       wants to subscribe to.
+        broker_addr (tuple[str, int], optional): Address of the broker the container shall connect to. Defaults to ("127.0.0.1", 5555).
+        codec (None | Codec, optional): The codec of the container
+
+    Returns:
+        RunWithMQTTManager: _description_
+    """
+    return RunWithMQTTManager(num, agents, broker_addr=broker_addr, codec=codec)
+
+
+class ComposedAgent(RoleAgent):
+    pass
+
+
+def agent_composed_of(*roles: Role, register_in: None | Container) -> ComposedAgent:
+    """Create an agent composed of the given `roles`. If a container is provided,
+    the created agent is automatically registered with the container `register_in`.
+
+    Args:
+        *roles Role: The roles which are added to the agent
+        register_in (None | Container): container in which the created agent is registered,
+                                        if provided
+    """
+    agent = ComposedAgent()
+    for role in roles:
+        agent.add_role(role)
+    if register_in is not None:
+        register_in.register_agent(agent)
+    return agent
+
+
+class PrintingAgent(Agent):
+    def handle_message(self, content, meta: dict[str, Any]):
+        logging.info("Received: %s with %s", content, meta)
+
+
+def sender_addr(meta: dict) -> AgentAddress:
+    """Extract the sender_addr from the meta dict.
+
+    Args:
+        meta (dict): the meta you received
+
+    Returns:
+        AgentAddress: Extracted agent address to be used for replying to messages
+    """
+    # convert decoded addr list to tuple for hashability
+    return AgentAddress(
+        tuple(meta["sender_addr"])
+        if isinstance(meta["sender_addr"], list)
+        else meta["sender_addr"],
+        meta["sender_id"],
+    )
+
+
+def addr(protocol_part: Any, aid: str) -> AgentAddress:
```

Comment:
If anything changes with the naming on the previous comment on `AgentAddress` and `addr.addr` change would have to be propagated to this function too.